WEB-INTEGRATION.md

```diff
@@ -0,0 +1,742 @@
+exception was thrown.
+
+
+## Promises / async web APIs
+
+The `unhandledrejection` event on a window or worker global object is fired
+whenever a promise resolves with a rejection, without having a rejection handler
+registered. Exposing the context in which the promise was rejected is useful for
+the same reasons as for runtime script errors. Therefore,
+[`PromiseRejectionEvent`](https://html.spec.whatwg.org/multipage/webappapis.html#promiserejectionevent)
+will have a `rejectionSnapshot` property reflecting this context.
+
+In the case of promises created by web APIs, it seems the rejection context
+would always be the same as the context when the promise is created. According
+to the categories in the [“Writing Promise-Using
+Specifications”](https://w3ctag.github.io/promises-guide/) guide:
+- For one-and-done operations, the rejection-time context of the returned
+  promise should be the context when the web API that returns it was called.
+- For one-time “events”, the promise is created at the same time as an async
+  operation is started which will eventually resolve it, so the context would
+  flow from creation to rejection. E.g, for the
+  [`loaded`](https://drafts.csswg.org/css-font-loading-3/#dom-fontface-loaded)
+  property of a [`FontFace`](https://drafts.csswg.org/css-font-loading-3/#fontface)
+  instance, creating the `FontFace` instance causes both the promise creation
+  and the loading of the font.
+- More general state transitions are similar to one-time “events” in that each
+  promise is created at the same time as an async operation starts that will
+  resolve it.
+
+## Cross-document navigations
+
+When a cross-document navigation happens, even if it is same-origin, the context
+will be reset such that document load and tasks that directly flow from it
+(including execution of classic scripts found during parsing) run with the
+initial AsyncContext snapshot. The initial AsyncContext Snapshot will be
+provided by the host environment (i.e. by the HTML spec, for the web), and it
+should either be empty, or only have mappings from spec-defined
+`AsyncContext.Variable`s (see “Using AsyncContext from web specs” below).
+
+## Module evaluation
+
+When you import a JS module multiple times, it will only be fetched and
+evaluated once. Since module evaluation should not be racy (i.e. it should not
+depend on the order of various imports), the context should be reset so that
+module evaluation always runs with the initial AsyncContext snapshot. Inside of
+`ShadowRealm`s, the AsyncContext snapshot for all module evaluations will be the
+snapshot which was active when the `ShadowRealm` was created.
+
+# Editorial aspects of AsyncContext integration in web specifications
+
+An agent always has an associated AsyncContext mapping, in its
+`[[AsyncContextMapping]]` field. When the agent is created, this mapping will be
+set to an HTML-provided initial state, but JS user code can change it in a
+strictly scoped way.
```

Comment:
I *think* I understand why this is agent-wide instead of global-wide, but extra explanation (e.g. of what would go wrong if it were global-wide) would be helpful.